modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java

```diff
@@ -0,0 +1,537 @@
+                connectTimeout, connectionsPerNodeRecovery, connectionsPerNodeBulk, connectionsPerNodeReg, connectionsPerNodeState,
+                connectionsPerNodePing, receivePredictorMin, receivePredictorMax);
+        }
+
+        final ThreadFactory workerFactory = daemonThreadFactory(this.settings, HTTP_SERVER_WORKER_THREAD_NAME_PREFIX, name);
+
+        final ServerBootstrap serverBootstrap = new ServerBootstrap();
+
+        if (TCP_BLOCKING_SERVER.get(settings)) {
+            serverBootstrap.group(new OioEventLoopGroup(workerCount, workerFactory));
+            serverBootstrap.channel(OioServerSocketChannel.class);
+        } else {
+            serverBootstrap.group(new NioEventLoopGroup(workerCount, workerFactory));
+            serverBootstrap.channel(NioServerSocketChannel.class);
+        }
+
+        serverBootstrap.childHandler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) throws Exception {
+                ch.pipeline().addLast("open_channels", Netty4Transport.this.serverOpenChannels);
+                ch.pipeline().addLast("size", new Netty4SizeHeaderFrameDecoder(Netty4Transport.this));
+                ch.pipeline().addLast("dispatcher", new Netty4MessageChannelHandler(Netty4Transport.this, name));
+            }
+        });
+
+        serverBootstrap.childOption(ChannelOption.TCP_NODELAY, TCP_NO_DELAY.get(settings));
+        serverBootstrap.childOption(ChannelOption.SO_KEEPALIVE, TCP_KEEP_ALIVE.get(settings));
+
+        final ByteSizeValue tcpSendBufferSize = TCP_SEND_BUFFER_SIZE.getDefault(settings);
+        if (tcpSendBufferSize != null && tcpSendBufferSize.bytes() > 0) {
+            serverBootstrap.childOption(ChannelOption.SO_SNDBUF, Math.toIntExact(tcpSendBufferSize.bytes()));
+        }
+
+        final ByteSizeValue tcpReceiveBufferSize = TCP_RECEIVE_BUFFER_SIZE.getDefault(settings);
+        if (tcpReceiveBufferSize != null && tcpReceiveBufferSize.bytes() > 0) {
+            serverBootstrap.childOption(ChannelOption.SO_RCVBUF, Math.toIntExact(tcpReceiveBufferSize.bytesAsInt()));
+        }
+
+        serverBootstrap.option(ChannelOption.RCVBUF_ALLOCATOR, recvByteBufAllocator);
+        serverBootstrap.childOption(ChannelOption.RCVBUF_ALLOCATOR, recvByteBufAllocator);
+
+        final boolean reuseAddress = TCP_REUSE_ADDRESS.get(settings);
+        serverBootstrap.option(ChannelOption.SO_REUSEADDR, reuseAddress);
+        serverBootstrap.childOption(ChannelOption.SO_REUSEADDR, reuseAddress);
+
+        serverBootstrap.validate();
+
+        serverBootstraps.put(name, serverBootstrap);
+    }
+
+    protected final void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        Throwable t = cause;
+        while (t != null) {
+            if (!t.getClass().getCanonicalName().startsWith("org.elasticsearch")) {
```

Comment:
Yeah, I don't like that we have to do this at all, but exceptions that get thrown from TcpTransport end up getting caught and wrapped inside of Netty so we have to unwrap because going back down to `TcpTransport#onException`. I pushed 8304d68b839197894fb4e9016418c645da7c4299.